framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -2678,40 +2679,78 @@ protected void checkTypecastSafety(TypeCastTree typeCastTree) {
             }
             reported = true; // don't issue cast unsafe warning.
         }
-
         // Don't call TypeHierarchy#isSubtype(exprType, castType) because the underlying Java types
         // will not be in the correct subtyping relationship if this is a downcast.
-        if (!reported && !isTypeCastSafe(castType, exprType)) {
-            checker.reportWarning(
-                    typeCastTree, "cast.unsafe", exprType.toString(true), castType.toString(true));
+        if (!reported) {
+            TypecastKind castResult = isTypeCastSafe(castType, exprType);
+
+            if (castResult == TypecastKind.WARNING) {
+                checker.reportWarning(
+                        typeCastTree,
+                        "cast.unsafe",
+                        exprType.toString(true),
+                        castType.toString(true));
+            } else if (castResult == TypecastKind.ERROR) {
+                checker.reportError(
+                        typeCastTree,
+                        "cast.incomparable",
+                        exprType.toString(true),
+                        castType.toString(true));
+            }
         }
     }
 
     /**
-     * Returns true if the cast is safe.
-     *
-     * <p>Only primary qualifiers are checked unless the command line option "checkCastElementType"
-     * is supplied.
+     * This method returns TypecastKind.SAFE when the typecast is an upcast or a statically
+     * verifiable downcast. Returns TypecastKind.WARNING and ERROR for those downcasts which cannot
+     * be statically verified or some incomparable casts.
      *
      * @param castType annotated type of the cast
      * @param exprType annotated type of the casted expression
-     * @return true if the type cast is safe, false otherwise
+     * @return TypecastKind.SAFE if the typecast is safe, error or warning otherwise
      */
-    protected boolean isTypeCastSafe(AnnotatedTypeMirror castType, AnnotatedTypeMirror exprType) {
-        TypeKind castTypeKind = castType.getKind();
-        if (castTypeKind == TypeKind.DECLARED) {
-            // Don't issue an error if the annotations are equivalent to the qualifier upper bound
-            // of the type.
-            AnnotatedDeclaredType castDeclared = (AnnotatedDeclaredType) castType;
-            AnnotationMirrorSet bounds =
-                    atypeFactory.getTypeDeclarationBounds(castDeclared.getUnderlyingType());
+    protected TypecastKind isTypeCastSafe(
```

Comment:
A method that starts with `is` should usually return a boolean. Can you think of a more appropriate name?
We are already breaking all existing clients, so we might as well pick a better name.